### DISCUSSION TITLE
能否支持音乐播放器自定义歌曲源，比如引用站点附件的mp3文件

### DISCUSSION THREAD

**Discussion by IshuaWang** [hcjike/halo-theme-dream2.0-plus#37] [General]
目前只能引用在线的歌单，但是有的音乐在不同的平台，或者想上传自己制作的音频，目前好像没法做到。

**Comment by mjsoftking** [comment]
使用自定义侧边栏，自行编写html、js、css代码

**Reply by IshuaWang** [reply]
我试了下好像没成功。我就是自己定义了个music.js文件,放在assets/js目录：
```
const ap = new APlayer({
    container: document.getElementById('aplayer'),
    fixed: true,
    autoplay: true,
    mini: false,
    loop: 'all',
    order: 'random',
    preload: 'auto',
    mutex: true,
    volume: 0.7,
    lrcType: 0,
    listFolded: true,
    listMaxHeight: 60,
    audio: [
      {
        name: '尘世美',
        artist: '许嵩',
        url: '/upload/%E5%B0%98%E4%B8%96%E7%BE%8E%20-%20%E8%AE%B8%E5%B5%A9.mp3',
        cover: '',
      },
      {
        name: '夏天',
        artist: '欧阳娜娜',
        url: '/upload/%E5%A4%8F%E5%A4%A9%20-%20%E6%AC%A7%E9%98%B3%E5%A8%9C%E5%A8%9C.mp3',
        cover: '',
      },
    ]
  });
```
然后再自定义模块的内容里填入：
```
<div style="min-height: 90px;margin-top: -0.5rem;">
 <div id="aplayer">播放器</div>
</div>
  <link rel="stylesheet" th:href="@{/assets/lib/aplayer@1.10.1/APlayer.min.css}">
  <script defer th:src="@{/assets/lib/aplayer@1.10.1/APlayer.min.js}"></script>
<script defer th:src="@{/assets/js/music.js}"></script>
```
不知道为什么播放器这三个字能出来，但是播放器界面就是没有，能看看问题出在哪吗？

**Comment by mjsoftking** [comment]
```
<mew-music url="https://blog.wangyx.com/upload/%E5%B0%98%E4%B8%96%E7%BE%8E%20-%20%E8%AE%B8%E5%B5%A9.mp3" autoplay="true" name="演示音乐" artist="演示歌手" cover="/upload/thumbnails/2024/w800/th.jpg"></mew-music>
```

目前只支持单个音乐
name：歌曲名
artist：歌手名
cover：封面
autoplay：自动播放，受浏览器限制无法自动播放

**Reply by IshuaWang** [reply]
> ```
> <mew-music url="https://blog.wangyx.com/upload/%E5%B0%98%E4%B8%96%E7%BE%8E%20-%20%E8%AE%B8%E5%B5%A9.mp3" autoplay="true" name="演示音乐" artist="演示歌手" cover="/upload/thumbnails/2024/w800/th.jpg"></mew-music>
> ```
> 
> 目前只支持单个音乐 name：歌曲名 artist：歌手名 cover：封面 autoplay：自动播放，受浏览器限制无法自动播放

怎么做才能支持多个音乐啊，我之前是参照的这个博客，它是直接用aplayer做的，我看他的博客确实实现了这个效果
https://jinqq.gitlab.io/2024/12/16/%E5%A6%82%E4%BD%95%E6%90%AD%E5%BB%BA%E4%B8%AA%E4%BA%BA%E7%BD%91%E9%A1%B5%EF%BC%88%E5%85%AB%EF%BC%89%E2%80%94%E2%80%94%E9%9F%B3%E4%B9%90%E6%92%AD%E6%94%BE%E5%99%A8/
但是我在halo上没试成功

**Comment by mjsoftking** [comment]
https://github.com/zsjy/halo-theme-dream2.0-plus/releases/tag/1.3.3.beta6

**Comment by mjsoftking** [comment]
```
<mew-music listFolded="true" music-list='[{"name":"1", "url":"https://blog.wangyx.com/upload/%E5%B0%98%E4%B8%96%E7%BE%8E%20-%20%E8%AE%B8%E5%B5%A9.mp3","artist":"1","cover":"/upload/thumbnails/2024/w800/th.jpg","lrc":""},{"name":"2", "url":"https://blog.wangyx.com/upload/%E5%B0%98%E4%B8%96%E7%BE%8E%20-%20%E8%AE%B8%E5%B5%A9.mp3","artist":"2","cover":"/upload/thumbnails/2024/w800/th.jpg","lrc":""}]'></mew-music>
```
music-list: 传递json数据，必须是数组且包含示例中的所有参数

**Comment by mjsoftking** [comment]
### 组件属性
- ‌autoplay‌：是否自动播放，默认为`false`，受限一般无效。
- ‌theme‌：播放器主题颜色，默认为 `主题色`。
- ‌loop‌：循环模式，可选值包括`'all'、'one'、'none'`，默认:` 'all'`。
- ‌volume‌：默认音量，默认为 `0.7`。
- ‌listFolded‌：列表是否默认折叠，默认为`false`。
- ‌listMaxHeight‌：列表最大高度，默认为`450`。
- ‌lrcType‌：歌词文件形式，可选值包括`1（直接复制歌词）、2（HTML形式）、3（数据库获取）`，默认：`3`。
- music-list：播放器中的歌曲，`json数组`
#### music-list 属性
- 值必须是json数组，即：`[]`
- name：歌曲名
- url：歌曲链接
- artist：歌手名
- cover：封面
- lrc：歌词

**Reply by IshuaWang** [reply]
这个可以关闭歌词吗？不设置的话老是跳出歌词不存在的错误，还有我设置mini好像没起到作用不知道是为什么

**Reply by mjsoftking** [reply]
设置为空字符串就行，或者用这个
https://github.com/zsjy/halo-theme-dream2.0-plus/releases/tag/1.3.3.beta8
加了默认值处理，同时改为使用内置的js、css，避免免费cdn加载慢

**Reply by IshuaWang** [reply]
我发现了个现象，就是第一次打开网站的时候播放器不会自动播放，但是点击其他页面的时候就能自动播放了，这是为什么啊？按照这个原理打开网页之后再定向到首页是不是可以实现自动播放的效果？

**Reply by mjsoftking** [reply]
你查那个错误就知道了，浏览器控制，需要用户进行交互才能执行自动播放类的行为

**Reply by IshuaWang** [reply]
> 你查那个错误就知道了，浏览器控制，需要用户进行交互才能执行自动播放类的行为

现在的版本我好像切换页面也没法播放，而且切换页面会导致歌曲变了，有什么办法能让切换页面前后播放中的歌曲不变吗？

**Reply by mjsoftking** [reply]
开启pjax，否则没办法，pjax和halo评论组建不兼容存在问题，开启pjax需要选择web评论组建

**Reply by IshuaWang** [reply]
> 你查那个错误就知道了，浏览器控制，需要用户进行交互才能执行自动播放类的行为

用户交互是必须执行某种操作吗？还是说滚动滚轮，鼠标左击都可以？如果是这样是不是可以监听事件来播放。目前就算开启pjax，还是会把播放暂停

**Reply by mjsoftking** [reply]
应该是点击页面上的什么按钮之类的

**Reply by IshuaWang** [reply]
出现了这两个错误不知道是怎么回事？
![image](https://github.com/user-attachments/assets/222d6aa7-8802-4c92-883d-7aaf80d5beb6)
而且我突然觉得有的地方网页切换很慢，不知道是不是这个造成的

**Comment by mjsoftking** [comment]
![image](https://github.com/user-attachments/assets/02943a4b-4d08-40d4-aa89-7660ca711a49)
https://github.com/zsjy/halo-theme-dream2.0-plus/releases/tag/1.3.3.beta9

这2个文件大小并不大，所以只要服务器带宽不是很差不至于太慢

**Reply by IshuaWang** [reply]
我的服务器在新加坡，应该不是这个问题把

**Reply by mjsoftking** [reply]
那就很正常了，你本身访问就慢

**Comment by mjsoftking** [comment]
参数看这个：https://aplayer.js.org/#/zh-Hans/
之前是AI查到的可能解释存在错误，还有不要乱设置参数
![image](https://github.com/user-attachments/assets/648a9945-f435-42b5-9b60-25341341f894)
音频这些参数没有添加支持

**Reply by IshuaWang** [reply]
我只设置了name, artist和url，其他的应该不需要设置了把

**Reply by mjsoftking** [reply]
数据中只需要强制的字段只有url，其他的都有默认值

**Reply by IshuaWang** [reply]
目前如果不设置cover，好像播放器界面跳一下就没有了

**Reply by mjsoftking** [reply]
你设置了固定底部位置

**Reply by IshuaWang** [reply]
但是我加了cover之后用fixed好像就能出现，有点奇怪

**Comment by mjsoftking** [comment]
https://github.com/zsjy/halo-theme-dream2.0-plus/releases/tag/1.3.3.beta11

文档：https://www.hcjike.com/docs/halo-theme-dream2.0/1.3.3.bate/theme/sidebar-assembly#%E6%B7%BB%E5%8A%A0%E5%B7%A6%E4%B8%8B%E8%A7%92%E9%9F%B3%E4%B9%90%E6%92%AD%E6%94%BE%E5%99%A8

**Reply by IshuaWang** [reply]
这个内嵌html内容和Halo设置里的代码注入是一样的吗？
